compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,221 @@
+            return false;
+        };
+
+        let entry_block = &mut function.dfg[entry_block_id];
+
+        // If the entry block has instructions, we can't inline it (we need a terminator)
+        if !entry_block.instructions().is_empty() {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        let terminator = entry_block.take_terminator();
+        let TerminatorInstruction::Return { return_values, call_stack: _ } = terminator else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
+
+        // Sanity check: make sure all returned values are constant
+        if !return_values.iter().all(|value_id| function.dfg.is_constant(*value_id)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // Replace the instruction results with the constant values we got
+        let current_results = self.dfg.instruction_results(instruction_id).to_vec();
+        assert_eq!(return_values.len(), current_results.len());
+
+        for (current_result_id, return_value_id) in current_results.iter().zip(return_values) {
+            let new_return_value_id = function.copy_constant_to_function(return_value_id, self);
+            self.dfg.set_value_from_id(*current_result_id, new_return_value_id);
+        }
+
+        true
+    }
+
+    /// Copies a constant from this function to another one.
+    /// Though it might seem we can just take a value out of `self` and call `make_value` on `function`,
+    /// if the constant is an array the values will still keep pointing to `self`. So, this function
+    /// recursively copies the array values too.
+    fn copy_constant_to_function(&self, constant_id: ValueId, function: &mut Function) -> ValueId {
+        if let Some((constant, typ)) = self.dfg.get_numeric_constant_with_type(constant_id) {
+            function.dfg.make_constant(constant, typ)
+        } else if let Some((constants, typ)) = self.dfg.get_array_constant(constant_id) {
+            let new_constants = constants
+                .iter()
+                .map(|constant_id| self.copy_constant_to_function(*constant_id, function))
+                .collect();
+            function.dfg.make_array(new_constants, typ)
+        } else {
+            unreachable!("A constant should be either a numeric constant or an array constant")
+        }
+    }
+}
+
+/// Optimizes a function by running the same passes as `optimize_into_acir`
+/// after the `inline_const_brillig_calls` pass.
+/// The function is changed to be an ACIR function so the function can potentially
+/// be optimized into a single return terminator.
+fn optimize(
+    mut function: Function,
+    inliner_aggressiveness: i64,
+    error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+) -> Result<Function, RuntimeError> {
+    function.set_runtime(RuntimeType::Acir(InlineType::InlineAlways));
```

Comment:
Why do we need to set this particular runtime for this function? What would `Acir(Inline)` or `Brillig(InlineAlways)` do? And doesn't any function called by this one need to have the same thing done to it?

Reply:
I don't know much about this, but I saw in the code that some optimizations don't run for brillig code (for example loop unrolling). I also asked Tom in which mode we should optimize and he said Acir. Maybe @TomAFrench or someone else knows what can or can't be optimized in acir/brillig?

Reply:
I suggested treating the code as if it were to be ACIR as we don't unroll loops in brillig code and without loop unrolling we're not going to be able to simplify a function down to nothing generally.

This does result in some issues where we can't deal with brillig-only language features such as break however. Perhaps a better solution to this problem is to defer the handling of constant-input brillig functions until after brillig gen. We can then execute the brillig functions directly (so we get the results without needing to be able to fully optimize them in SSA) and then feed those back into the SSA form of the program to do further optimizations.

The alternative to this is to execute the SSA directly but that adds yet another execution environment so I'm reluctant to do so.

Reply:
Thanks, it would be great to preserve this explanation as a comment. 

Some followup questions:
* Is this setting transitive, ie. what if this function calls another function that is a brillig function with a loop?
* What happens if we mark a brillig function with `break` as `Acir(InlineAlways)`? I suspect that's why something further down was turned into `Result`, but I don't see it in the `run_passes` returning an error, and I remember seeing errors where the compiler couldn't figure out the loop bounds at compile time in constrained code.

Reply:
Oh I see that we only pass this function to the SSA in `let ssa = Ssa::new(vec![function],`. 

What I don't fully understand is when [should_inline_call](https://github.com/noir-lang/noir/blob/b74b4eaf7acf6becd0f98ab25d0d5ccde3affebe/compiler/noirc_evaluator/src/ssa/opt/inlining.rs#L638-L639) makes a decision about [calls made by this function](https://github.com/noir-lang/noir/blob/b74b4eaf7acf6becd0f98ab25d0d5ccde3affebe/compiler/noirc_evaluator/src/ssa/opt/inlining.rs#L607-L610) it looks for the definition of the callee in the SSA, and it won't find it. It looks like it would panic there on `let function = &ssa.functions[&called_func_id];` 🤔 

And that function also answers the question of what would happen if the called function is brillig: it's not inlined because the caller is marked as `Acir`.

Reply:
> What I don't fully understand is when [should_inline_call](https://github.com/noir-lang/noir/blob/b74b4eaf7acf6becd0f98ab25d0d5ccde3affebe/compiler/noirc_evaluator/src/ssa/opt/inlining.rs#L638-L639) makes a decision about [calls made by this function](https://github.com/noir-lang/noir/blob/b74b4eaf7acf6becd0f98ab25d0d5ccde3affebe/compiler/noirc_evaluator/src/ssa/opt/inlining.rs#L607-L610) it looks for the definition of the callee in the SSA, and it won't find it

That's why inlining is not performed when optimizing this function. Or, put another way, inlining was already done before we start optimizing this function. And the other inline pass (for no predicates) is skipped for this reason too.

Reply:
Ah, okay, I thought that `false` would have something to do with it. Well I guess if it wasn't correct it would have panicked already 👍